src/actions/sale/getSaleData.ts

```diff
@@ -0,0 +1,41 @@
+import { ethers } from "ethers";
+import { SalePhases } from "../../constants";
+import { Sale } from "../../contracts/types";
+import { SaleConfiguration, SaleData, SalePhase } from "../../types";
+
+export const getSaleData = async (
+  contract: Sale,
+): Promise<SaleData> => {
+  const salePhase: SalePhase =
+    (await contract.salePhase()) as SalePhase;
+  const salePhaseName = SalePhases[salePhase];
+  const saleStartTime = (await contract.saleStartBlockTimestamp()).toNumber();
```

Comment:
we, unfortunately, can't use Number type at all with smart contracts. it is possible sometimes, but should not be used, since for other cases it can trigger problems. in Solidity we deal with very large numbers mostly, and they are not supported by "Number".
when we work with Solidity contracts (and it would be better everywhere else also), any number we use should be a BigNumber or BigNumber.toString().

Reply:
Makes sense, but I think timestamps are a case where use of number is valid number is not valid.  In the specific case of time, it is measured on solidity in `seconds`. 

From the JS spec:
_Time is measured in ECMAScript in milliseconds since 01 January, 1970 UTC. In time values leap seconds are ignored. It is assumed that there are exactly 86,400,000 milliseconds per day. ECMAScript Number values can represent all integers from –9,007,199,254,740,992 to 9,007,199,254,740,992; this range suffices to measure times to millisecond precision for any instant that is within approximately 285,616 years, either forward or backward, from 01 January, 1970 UTC._

I'll review the other data values though and make sure I'm not misusing number anywhere

Reply:
Looking at the other values that convert to number, I think a similar case can be made:
`saleCounter` - safe until approx 9,007,199,254,740,992 sales
`amountForSale` - safe, we can maybe revisit if we're ever doing anything wacky with fractionalization
`startingMetadataIndex` - part a of a folder group, will cap out at the amount for sale (thousands)

Reply:
I personally suggest to use more clear data type.
@Whytecrowe is right, we are mainly using `uint256` in SC, it will represent as BigNumber in ethers.js, but sometimes, ethers.js automatically convert some values into the number if their limit is not too big, you can check those data types in the typechain files.
i.e. `const timestamp: number = await contract.saleStartBlockTimestamp()` (no need to convert to number)
I think token amount, token price, id, etc there are strongly compatible with BigNumber, but better to use number for timestamp, current phase, counter, etc.
In the frontend perspective, number is more useful for these fields, 
i.e. `const startTime = new Date((await contract.saleStartBlockTimestamp()) * 1000); console.log(startTime.toLocalString())`
What do you think? @Whytecrowe @RyanCross

Reply:
Yeah. I think the best way for me personally is to just use BN everywhere on our side. Then we can let the front team know that this is the standard for us and they can adapt all of their code to convert them into whatever type they need. for them it might be easy as well, knowing that everything will come in one type.
I am personally not a big fan of having different types for essentially the same type, cause often it's just tedious to figure it out every time you are writing code (we have enough of other types to worry about, including smart contracts). which type is this number? is it a bn is it not? is this big enough to be a bn or should we use "number" type? all this is just extra time and distraction from the main problem you are trying to solve and gets very annoying with time. but maybe it's just me :) not a huge problem though.

If it's a problem for the frontend guys, and it's hard for them to convert from BN every time, then we can adapt.

Reply:
@Whytecrowe In most cases we should be using BigNumber or the `toString()` function from it but these are small regular numbers and Ryan's case is valid. In terms of future tech debt we don't have to back and modify all the existing work like you claim if we have `BigNumberish` types like we do here or elsewhere.